lib/spack/spack/util/compression.py

```diff
@@ -248,70 +280,357 @@ def _7zip(archive_file):
+            # as is absolutely neccesary for largest archive header (tar)
+            max_size = TarFileType.OFFSET + TarFileType.header_size()
+            return io.BytesIO(
+                initial_bytes=lzma.LZMADecompressor().decompress(
+                    stream.read(max_size), max_length=max_size
+                )
+            )
+        return None
+
+
+class TarFileType(FileTypeInterface):
+    OFFSET = 257
+    _MAGIC_NUMBER_GNU = b"ustar  \0"
+    _MAGIC_NUMBER_POSIX = b"ustar\x0000"
+    extension = "tar"
+
+    @staticmethod
+    def name():
+        return "tar archive"
+
+
+class ZipFleType(FileTypeInterface):
+    _MAGIC_NUMBER = b"PK\003\004"
+    extension = "zip"
+
+    @staticmethod
+    def name():
+        return "Zip archive data"
+
+
+# collection of valid Spack recognized archive and compression
+# file type identifier classes.
+VALID_FILETYPES = [
+    BZipFileType,
+    ZCompressedFileType,
+    GZipFileType,
+    LzmaFileType,
+    TarFileType,
+    ZipFleType,
+]
+
+
+def extension_from_stream(stream, decompress=False):
+    """Return extension represented by stream corresponding to archive file
+    If stream does not represent an archive type recongized by Spack
+    (see `spack.util.compression.ALLOWED_ARCHIVE_TYPES`) method will return None
+
+    Extension type is derived by searching for identifying bytes
+    in file stream.
+
+    Args:
+        stream : stream representing a file on system
+        decompress (bool) : if True, compressed files are checked
+                            for archive types beneath compression i.e. tar.gz
+                            default is False, otherwise, return top level type i.e. gz
+
+    Return:
+        A string represting corresponding archive extension
+            or None as relevant.
+
+    """
+    for arc_type in VALID_FILETYPES:
+        if arc_type.is_file_of_type(stream):
+            suffix_ext = arc_type.extension
+            prefix_ext = ""
+            if arc_type.compressed and decompress:
+                # stream represents compressed file
+                # get decompressed stream (if possible)
+                decomp_stream = arc_type.decomp_in_memory(stream)
+                prefix_ext = extension_from_stream(decomp_stream, decompress=decompress)
+                if not prefix_ext:
+                    # We were unable to decompress or unable to derive
+                    # a nested extension from decompressed file.
+                    # Try to use filename parsing to check for
+                    # potential nested extensions if there are any
+                    tty.debug(
+                        "Cannot derive file extension from magic number;"
+                        " falling back to regex path parsing."
+                    )
+                    return extension_from_path(stream.name)
+            tty.debug("File extension successfully dervived by magic number.")
```

Comment:
Sorry it just occurred to me it would be good to also mention the type that was detected (so e.g. spack debug output would say "based on magic numbers I think it's a .tar.gz" - so if we run into trouble later, a user can manually check and confirm).

Also typo: dervived